src/3d/qgs3dmodelcache.h

```diff
@@ -0,0 +1,112 @@
+                         qgs3dmodelcache.h
+                         ---------------
+    begin                : July 2020
+    copyright            : (C) 2020 by Mathieu Pellerin
+    email                : nirvn dot asia at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGS3DMODELCACHE_H
+#define QGS3DMODELCACHE_H
+
+#include "qgis_3d.h"
+#include "qgsabstractcontentcache.h"
+
+#include <Qt3DRender/QSceneLoader>
+#include <Qt3DRender/QMesh>
+
+
+///@cond PRIVATE
+
+/**
+ * \ingroup 3d
+ * \class Qgs3DModelCacheEntry
+ * An entry for a Qgs3DModelCache, representing a 3D model filename
+ * \since QGIS 3.16
+ */
+class _3D_EXPORT Qgs3DModelCacheEntry : public QgsAbstractContentCacheEntry
+{
+  public:
+
+    /**
+     * Constructor for Qgs3DModelCacheEntry, corresponding to the specified \a path.
+     */
+    Qgs3DModelCacheEntry( const QString &path ) ;
+
+    //! The 3D model temporary filename
+    QString temporaryFile;
+
+    int dataSize() const override;
+    void dump() const override;
+    bool isEqual( const QgsAbstractContentCacheEntry *other ) const override;
+
+};
+
+///@endcond
+
+/**
+ * \class Qgs3DModelCache
+ * \ingroup 3d
+ * A cache for 3D models.
+ *
+ * Qgs3DModelCache delivers pre-fetched 3D models.
+ *
+ * Qgs3DModelCache is not usually directly created, but rather accessed through
+ * QgsApplication::3dModelCache().
+ *
+ * \since QGIS 3.16
+*/
+class _3D_EXPORT Qgs3DModelCache : public QgsAbstractContentCache< Qgs3DModelCacheEntry >
+{
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Constructor for Qgs3DModelCache, with the specified \a parent object.
+     */
+    Qgs3DModelCache( QObject *parent SIP_TRANSFERTHIS = nullptr );
+
+    /**
+     * Returns a QMesh object with its source set to the specified \a path
+     *
+     * \a path may be a local file, remote (HTTP) url, or a base 64 encoded string (with a "base64:" prefix).
+     */
+    Qt3DRender::QMesh *pathAsMesh( const QString &path, bool blocking = false );
+
+    /**
+     * Returns a QSceneLoader object with its source set to the specified \a path
+     *
+     * \a path may be a local file, remote (HTTP) url, or a base 64 encoded string (with a "base64:" prefix).
+     */
+    Qt3DRender::QSceneLoader *pathAsSceneLoader( const QString &path, bool blocking = false );
```

Comment:
Hmm... to me whether a model is loaded as QMesh or as a QSceneLoader should be an implementation detail of the 3d point symbol implementation... could we have API that would just return URI and leave the rest to qgspoint3dsymbol_p.cpp ?

Reply:
@wonder-sk , initially, I thought we could actually cache the QMesh and QSceneLoader and actually gain some speed boost out of it. Turns out Qt doesn't allow copying of QMesh and QSCeneLoader objects at the moment. I'm trying to see if we can return copies of a cached QEntity. If that can't be done, I don't mind returning QUrl URIs, unless you think Qt will eventually allow copying of the above-mentioned objects, then it might make sense to keep things in its current form.

Reply:
I would be very cautious with caching/sharing of some Qt3D instances - some classes must not be even used multiple times as components of different entities. And especially with QSceneLoader-based point symbols I think we will need to change implementation eventually as it does not scale with the number of points. Currently for N points of a layer we create N QSceneLoader objects which is not optimal and I have seen it crash with higher number of points (at least on older version of Qt). The correct solution would be to load QSceneLoader once, traverse its internals for sub-entities (meshes with materials and transforms) and then apply instanced rendering like we do for simple 3d point shapes where the geometry is only stored in memory once, but rendered multiple times for single frame. Based on all this, my preference is not to expose QSceneLoader and just return URIs :-)